### PR TITLE
Add regression tests for client module

### LINKED ISSUE #263: Add TLS/QUIC configuration tests (certificate parsing & quinn endpoint settings)
Scope:
- Add unit and integration tests for certificate parsing/loading and Endpoint/Client configuration used by crusher (quinn::Endpoint / rustls-related conversion functions).

Acceptance criteria:
- Tests exercise success and failure paths for loading PEM chains, private keys, and CA certificates.
- Assert that keep-alive and transport settings are applied correctly to created quinn Client/Server configs (where they are converted from app settings).
- Confirm invalid keys/CAs produce clear error results (no panics) and that error messages are stable for regression testing.

Implementation notes:
- Add test fixtures under tests/fixtures or generate temp files with the tempfile/tempdir crates.
- Use small self-signed certs or rustls test helpers to avoid external dependencies.
- Keep the tests deterministic and avoid network I/O where possible (focus on parsing/config conversion).

### PR DESCRIPTION
**Pull request by JonghoKim-jj**
Closes #263 

### Objective

This PR adds regression tests for the `client` module to verify current behavior and prepare for future security updates.

### What this PR introduces

**1. Scenario-driven Regression Coverage**
A comprehensive set of test cases mapping various TLS certificate states and edge cases.
This ensures that the module's behavior—especially regarding certificate bundle handling—is deterministic and verified.

**2. Tests as "Executable Specifications"**
We are introducing tests that define not just *current* behavior, but *desired* behavior.

* **Regression Tests:** Verify the current implementation works as expected.
* **Pending Requirements:** A subset of tests is written *ahead of implementation* to codify our alignment with **Modern Security Standards** (e.g., enforcing full-chain certificates and strict SAN validation).

**3. Zero CI Noise (Stability by Design)**
Tests for features/validations not yet implemented are marked with `#[ignore]` and prefixed with `pending_`.

* **Benefit:** This keeps the requirements visible in the codebase without causing CI failures.
* **Roadmap:** These tests serve as a clear "todo list" for future PRs, allowing us to enable them incrementally as we harden the implementation.

### Context & Rationale: Why is the test suite so large?

While `client.rs` is currently concise (~90 lines), it manages **security-sensitive invariants** (TLS).
The current implementation has minimal validation, which increases the risk of silent misconfiguration.
We invested in a comprehensive test suite (1000+ lines) to:

* **Pin down invariants:**
Lock in correct behavior before refactoring or expanding validation logic.
* **Expose hidden complexity:**
Make subtle correctness rules (e.g., CA mismatches, chain ordering) explicit in code rather than relying on tribal knowledge.

### Implementation Notes: `rstest`

This suite is implemented using the [`rstest`](https://github.com/la10736/rstest) crate as a lightweight way to express fixtures and parameterized cases in Rust. We chose it for:

* **Scalability & Maintainability:**
It allows us to cover the combinatorial explosion of TLS edge cases with minimal boilerplate. Adding new scenarios becomes a data-entry task rather than writing new functions.
* **Ecosystem Trust:**
`rstest` is widely adopted in the Rust community (55M+ downloads on [Crates.io](https://crates.io/crates/rstest)).
* **Industry Compatibility:**
`rstest` integrates seamlessly with **`tokio::test`**, allowing us to write parameterized async tests without friction.
Additionally, [`googletest-rust`](https://github.com/google/googletest-rust) explicitly mentions interoperability with `rstest`, confirming it fits well with professional, async-heavy workflows.

### Future Work

* **Validation Strategy:** Currently, the `client` module lacks specific validation functions. We need to discuss and design a proper validation layer to enforce the security rules defined in these tests.
* **Integration Testing:** This PR focuses on unit-level invariants. Subsequent PRs will add integration tests to validate end-to-end client connection flows.
* **Enabling Pending Tests:** We will progressively implement strict validation logic (Full-chain/SAN) and enable the pending tests.

### CODE REVIEW

File: src/client.rs [hunk 1]
```diff
@@ -85,21 +91,1055 @@ pub(crate) fn config(certs: &Certs) -> Result<Endpoint> {
...
+        let endpoint = config(&certs).expect("Endpoint should be created successfully");
+        let local_addr = endpoint
+            .local_addr()
+            .expect("Endpoint should have local address");
+        assert!(local_addr.port() > 0, "Endpoint should be bound to a port");
+    }
+
+    // =========================================================================
+    // Handshake Tests
+    // =========================================================================
```
[the remark this answers is not in the document]
Comment (JonghoKim-jj) [reply src/client.rs]: Thanks for the feedback. I see your point regarding the module's responsibility.

However, I'd like to clarify that `client.rs` is not responsible solely for certificate parsing; it also exports the `config()` function, which constructs the final `quinn::Endpoint` with a specific TLS configuration.

The intention behind these tests is to serve as regression tests for the `config()` logic itself. We need to verify that the generated Endpoint actually works with the provided certificate chains, and some issues (like chain ordering, missing SANs, or trust validation) can only be detected during an actual TLS handshake.

Since this crate (`crusher`) is structured as a binary crate without a `lib.rs`, moving those tests to the `tests/` directory would require exposing internal modules or restructuring the project. Therefore, I placed them here to ensure the client configuration is validated within the module's scope.

Do you think we should keep this validation here to ensure reliability, or would you still prefer a different approach given the structure?

**Comment by JonghoKim-jj** [comment]
> Since we should handle changes to the production code with great caution while adding tests, could you extract production code changes into a separate issue and handle it in a separate PR please?

Done.
I forgot removing them. Thanks for the feedback!

**Comment by JonghoKim-jj** [comment]
@sophie-cluml I changed the visibility of `REQUIRED_GIGANTO_VERSION` to `pub(crate)` to use it within the handshake test logic. However, this change caused the code coverage to drop, resulting in a CI failure.

Should I add a trivial test case verifying this constant to satisfy the checker, or would you prefer adjusting the CI threshold to allow for small drops?

**Review by sophie-cluml — COMMENTED** [review]
I noticed there are quite a few `#[ignore]` tests in this PR. Since they take up a significant number of lines, it can be distracting for reviewers to review code that isn't intended to be active yet. It also looks like some fixtures are only used by these ignored tests.

While there are several ways to handle this, I think the cleanest approach would be to remove the ignored tests and their associated fixtures from this PR entirely and track them in a separate issue. If that’s not feasible, completing the tests now or moving them into a separate commit would also work.

File: src/client.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (JonghoKim-jj) [reply src/client.rs]: Thanks for your feedback!
I will split common logic.

**Comment by JonghoKim-jj** [comment]
@pott-cho @sophie-cluml 

Following the review in https://github.com/aicers/crusher/pull/274#pullrequestreview-3654241065, I initially removed all functions and fixtures prefixed with `pending_`.

However, after giving it some more thought, I've changed my mind. 
I believe we should restore the fixtures for the handshake test cases.

The reasons are as follows:

- `crusher` currently lacks validation logic, and there has been no discussion regarding which module should handle it.
- We rely entirely on crates like `rustls` for TLS certificate verification.
- If the handshake succeeds with an invalid certificate, it is a "**BUG**". 
Although the actual fix (validation logic) will be addressed in a separate Issue/PR, we must keep these tests here to ensure the bug is detected. (and does not fail silently)

Does this sound reasonable to you?

**Comment by pott-cho** [comment]
@JonghoKim-jj 
Our current goal is test automation. If a pending test cannot be validated during the CI process, it would be better not to add it.
Additionally, rather than writing test code in advance for code that will be implemented later, it would be more appropriate to write the test code together when implementing that code in a separate issue.

**Comment by JonghoKim-jj** [comment]
@sophie-cluml I've created 2 issues: #279 #280

**Comment by JonghoKim-jj** [comment]
Based on comment https://github.com/aicers/crusher/pull/274#pullrequestreview-3675316592,
I made few changes:

- Renamed struct: `CertHintForServer` -> `ServerCertsSetup`
- Renamed field: `fullchain_root_cert` -> `client_trust_root`
- Renamed field: `first_ca_cert`, `first_ca_key_pair` -> `server_issuer_cert`, `server_issuer_key`
- Renamed variables and parameters: `hint` -> `cert_setup`
- Updated comments to accurately describe the fields.
